src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/VSMSBuildNuGetProjectSystem.cs

```diff
@@ -339,24 +339,36 @@ public virtual void AddReference(string referencePath)
         {
             if (referencePath == null)
             {
-                throw new ArgumentNullException("referencePath");
+                throw new ArgumentNullException(nameof(referencePath));
             }
 
-            string name = Path.GetFileNameWithoutExtension(referencePath);
+            var name = Path.GetFileNameWithoutExtension(referencePath);
+            var projectName = string.Empty;
+            var projectFullPath = string.Empty;
+            var assemblyFullPath = string.Empty;
+            var dteProjectFullName = string.Empty;
 
-            ThreadHelper.JoinableTaskFactory.Run(async delegate
-            {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            var resolvedToGac = false;
 
-                try
+            try
+            {
+                // Perform all DTE operations on the UI thread
+                ThreadHelper.JoinableTaskFactory.Run(async delegate
                 {
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                    // Read DTE properties from the UI thread
+                    projectFullPath = ProjectFullPath;
```

Comment:
nit: same work as the locals above to make it abundantly clear that we're lifting those references into the delegate